interpreter.py

```diff
@@ -0,0 +1,164 @@
+
+import sys
+
+from rpython.rlib import jit
+from llvm_wrapper import *
+from state import State
+
+def target(*args):
+    return main, None
+
+def jitpolicy(driver):
+    from rpython.jit.codewriter.policy import JitPolicy
+    return JitPolicy()
+
+jit_driver = jit.JitDriver (
+    greens = [],
+    reds = [])
+
+class Interpreter(object):
+
+    def __init__(self, global_state = None):
+        self.global_state = global_state
+
+    # only works for integers
+    def _get_args(self, args, state):
+        ''' Returns a list of arguments represented as integers '''
+
+        arg_vals = []
+        for arg in args:
+            if LLVMIsConstant(arg):
+                arg_vals.append(LLVMConstIntGetSExtValue(arg))
+            elif state.has_key(rffi.cast(rffi.INT, arg)):
+                arg_vals.append(state.get_variable(rffi.cast(rffi.INT, arg)))
+            elif self.global_state.has_key(rffi.cast(rffi.INT, arg)):
+                arg_vals.append(self.global_state.get_variable(rffi.cast(rffi.INT, arg)))
+            else:
+                print "error"
+        return arg_vals
+
+    def exec_operation(self, state, opcode, args=[]):
+        if opcode == LLVMRet:
+            return 0
+        elif opcode == LLVMAdd:
+            x, y = self._get_args(args, state)
+            return x + y
+        elif opcode == LLVMMul:
+            x, y = self._get_args(args, state)
+            return x * y
+        elif opcode == LLVMCall:
+            string_format_ref = LLVMGetOperand(args[0], 0)
+            string_format = None
+            # need to access @.str value
+            if state.has_key(rffi.cast(rffi.INT, string_format_ref)):
+                string_format = state.get_variable(rffi.cast(rffi.INT, string_format_ref))
+            elif self.global_state.has_key(rffi.cast(rffi.INT, string_format_ref)):
+                string_format = self.global_state.get_variable(rffi.cast(rffi.INT, string_format_ref))
+            else:
+                print "error"
+            fn_name = rffi.charp2str(LLVMGetValueName(args[-1]))
+            if fn_name == "printf":
+                # XXX assumes all printf arguments are integers
+                printf_args = []
+                for arg in args[1:-1]:
+                    if state.has_key(rffi.cast(rffi.INT, arg)):
+                        printf_args.append(state.get_variable(rffi.cast(rffi.INT, arg)))
+                    elif self.global_state.has_key(rffi.cast(rffi.INT, arg)):
+                        printf_args.append(self.global_state.get_variable(rffi.cast(rffi.INT, arg)))
+                    elif LLVMIsConstant(arg):
+                        printf_args.append(LLVMConstIntGetSExtValue(arg))
+                    else:
+                        # raise custom exception
+                        print "error"
+                print string_format % tuple(printf_args)
+            else:
+                raise NotImplementedError(fn_name)
+        elif opcode == LLVMAlloca:
+            pass
```

Comment:
If these are not impemented, print a message and raise `NotImplementedError`? Perhaps make a function to do that, e.g. `def exit_not_implemented(self, message): ...`